### PR TITLE
Fixed varsub on more complex vars

### PR DESCRIPTION
**Pull request by jilleJr**
- \[x] I've added a new note in the `CHANGELOG.md` file, according to docs:
  https://iver-wharf.github.io/#/development/changelogs/writing-changelogs

## Summary

- Fixed varsub to be less regex dependent
- Fixed varsub regex to allow dashes and other chars in variables
- Rewrote varsub to be more iterative, fixing edge cases.

## Motivation

The following did not work:

```yaml
image: ${my-variable}
```

Because the regex didn't account for dashes.

I changed to grab everything except the closing brace, to allow as complex variable names as possible.

Also, I fixed so `${   %foobar%   }` doesn't count as escaping. Only if the percent signs are just next to the braces do they count as escaping.

### CODE REVIEW

File: pkg/varsub/varsub_test.go [hunk 1]
```diff
@@ -148,7 +154,7 @@ func TestSubstitute(t *testing.T) {
 		{
 			name:  "simple text with escaped empty white signs 2",
 			value: "Foo ${ %\n \r%\n} bar",
-			want:  "Foo ${\n \r} bar",
+			want:  "Foo ${ %\n \r%\n} bar",
 		},
```
Comment (Alexamakans) [inline pkg/varsub/varsub_test.go:158]: Edge-case: `${lorem} ${%lorem%} ${lorem}`    ->    `ipsum ipsum ${lorem}`
Should be `ipsum ${lorem} ipsum`.

I.e, the test proposed below breaks, result is
```
expected: "Foo ipsum ${lorem} ipsum"
actual  : "Foo ipsum ipsum ${lorem}"
```

```suggestion
		},
		{
			name:  "unescaped variables don't mess with substitution of matching vars after",
			value: "Foo ${lorem} ${%lorem%} ${lorem}",
			want:  "Foo ipsum ${lorem} ipsum",
		},
```
Only happens if the result of the unescape matches one of the vars after it.
I.e., this is passes:
```golang
{
	name:  "unescaped variables don't mess with substitution of non-matching vars after",
	value: "Foo ${lorem} ${%foo-bar%} ${lorem}",
	want:  "Foo ipsum ${foo-bar} ipsum",
},
```

I'd be kind of okay with not handling this case, but it would be very nice to fix now so it doesn't bite us in the butt 😅

**Review by Alexamakans — CHANGES_REQUESTED** [review]
Looks good besides a wrong return.

Also found an edge-case you may or may not want to deal with 😉

**Comment by jilleJr** [comment]
> Looks good besides a wrong return.
> 
> Also found an edge-case you may or may not want to deal with wink

I ended up dealing with it. I rewrote the varsubbing so it's iterative in its substitutioning.

File: pkg/varsub/varsub.go [hunk 2]
```diff
@@ -34,39 +35,56 @@ func Substitute(value string, source Source) (any, error) {
...
+			return nil, ErrRecursiveLoop
+		}
+
+		v, ok := source.Lookup(match.Name)
+		if !ok {
+			sb.WriteString(match.FullMatch)
+			continue
+		}
+
+		var matchVal = v.Value
```
Comment (Alexamakans) [inline pkg/varsub/varsub.go:61]:
```suggestion
		matchVal := v.Value
```

**Review by Alexamakans — APPROVED** [review]
Nice, LGTM 👍🏻

File: pkg/varsub/varsub.go (hunk 2 above)
Comment (jilleJr) [reply pkg/varsub/varsub.go:61]: Oof, nice catch
